File: sketch_dec26a/tests/test_sketch.cpp
// Static tests for Arduino sketch and header
// This file uses plain C++ and does not require Arduino runtime.
// It checks presence of expected symbols by parsing the source files at build time.

#include <fstream>
#include <sstream>
#include <string>
#include <iostream>

static std::string read_file(const char* path) {
    std::ifstream in(path);
    if (!in) return std::string();
    std::ostringstream ss;
    ss << in.rdbuf();
    return ss.str();
}

int main() {
    const char* ino = "../sketch_dec26a.ino";
    const char* hdr = "../DHT.h";

    std::string s_ino = read_file(ino);
    std::string s_hdr = read_file(hdr);

    if (s_ino.empty()) {
        std::cerr << "ERROR: could not read " << ino << "\n";
        return 2;
    }
    if (s_hdr.empty()) {
        std::cerr << "ERROR: could not read " << hdr << "\n";
        return 2;
    }

    bool ok = true;

    // Check for setup() and loop()
    if (s_ino.find("void setup()") == std::string::npos) {
        std::cerr << "Missing setup() in sketch" << std::endl;
        ok = false;
    }
    if (s_ino.find("void loop()") == std::string::npos) {
        std::cerr << "Missing loop() in sketch" << std::endl;
        ok = false;
    }

    // Check for includes
    if (s_ino.find("#include <DHT.h>") == std::string::npos) {
        std::cerr << "Sketch does not include <DHT.h>" << std::endl;
        ok = false;
    }
    if (s_ino.find("#include <LiquidCrystal.h>") == std::string::npos) {
        std::cerr << "Sketch does not include <LiquidCrystal.h>" << std::endl;
        ok = false;
    }

    // Check header guard in DHT.h
    if (s_hdr.find("#ifndef DHT_H") == std::string::npos || s_hdr.find("#define DHT_H") == std::string::npos) {
        std::cerr << "DHT.h missing header guard DHT_H" << std::endl;
        ok = false;
    }

    // Check DHT class presence
    if (s_hdr.find("class DHT") == std::string::npos) {
        std::cerr << "DHT.h does not define class DHT" << std::endl;
        ok = false;
    }

    // Check for the DHT type macros/constants
    if (s_hdr.find("static const uint8_t DHT11") == std::string::npos) {
        std::cerr << "DHT11 constant missing in DHT.h" << std::endl;
        ok = false;
    }

    if (!ok) {
        std::cerr << "One or more static checks failed." << std::endl;
        return 1;
    }

    std::cout << "All static Arduino checks passed." << std::endl;
    return 0;
}

File: sketch_dec26a/tests/test_sketch.py
from pathlib import Path
import re

SKETCH_FILE = Path(__file__).resolve().parents[1] / 'sketch_dec26a.ino'
DHT_FILE = Path(__file__).resolve().parents[1] / 'DHT.h'


def test_sketch_has_setup_and_loop():
    text = SKETCH_FILE.read_text(encoding='utf-8')
    assert re.search(r'\bvoid\s+setup\s*\(', text), 'setup() not found in sketch'
    assert re.search(r'\bvoid\s+loop\s*\(', text), 'loop() not found in sketch'


def test_sketch_includes_dht_header():
    text = SKETCH_FILE.read_text(encoding='utf-8')
    assert ('#include "DHT.h"' in text) or ('#include <DHT.h>' in text), 'DHT.h not included in sketch'


def test_dht_header_guard_and_contents():
    text = DHT_FILE.read_text(encoding='utf-8')
    # Header guard
    assert re.search(r'#ifndef\s+\w+', text), 'No #ifndef header guard in DHT.h'
    assert re.search(r'#define\s+\w+', text), 'No #define header guard in DHT.h'
    assert re.search(r'#endif', text), 'No #endif in DHT.h'
    # Some DHT-related symbol or class
    assert ('class DHT' in text) or ('DHT(' in text) or ('#define DHT' in text) or ('read' in text), 'DHT.h seems to lack DHT-related content'
